lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2189,10 +2210,53 @@ balanceTx wrk pp timeTranslation partialTx = do
         utxoIndex
         (defaultChangeAddressGen argGenChange)
         changeState
-        partialTx
+        (over #extraUTxO (<> lookedUpUTxO) partialTx)
 
     return tx
   where
+    era = Write.recentEra @era
+
+    -- | 'assumeUTxOInEra era utxo' will convert the 'utxo' to the given era,
+    -- and throw an exception in 'IO' if the conversion fails.
+    --
+    -- It assumes the 'utxo' was queried from the node /after/ the 'era'.
+    assumeUTxOInEra
+        :: RecentEra era
+        -> MaybeInRecentEra Write.UTxO
+        -> IO (Write.UTxO era)
+    assumeUTxOInEra RecentEraConway (InRecentEraConway u)
+        = pure u
+    assumeUTxOInEra RecentEraBabbage (InRecentEraBabbage u)
+        = pure u
+    assumeUTxOInEra RecentEraBabbage (InRecentEraConway u)
+        -- The era changed to Conway between fetching the era and fetching the
+        -- utxo. Let's downgrade the utxo and continue.
+        = either (throwIO . ExceptionInvalidTxOutInEra) pure $ downgradeEra u
+    assumeUTxOInEra RecentEraConway (InRecentEraBabbage _)
+        -- It should be impossible for the node's tip to change from Conway back
+        -- to Babbage.
+        = impossibleRollback
+    assumeUTxOInEra _ InNonRecentEraAllegra
+        = impossibleRollback
+    assumeUTxOInEra _ InNonRecentEraAlonzo
+        = impossibleRollback
+    assumeUTxOInEra _ InNonRecentEraMary
+        = impossibleRollback
+    assumeUTxOInEra _ InNonRecentEraShelley
+        = impossibleRollback
+    assumeUTxOInEra _ InNonRecentEraByron
+        = impossibleRollback
+
+    impossibleRollback = error "assumeUTxOInEra: era should not roll back"
+
+    downgradeEra
+        :: Write.UTxO ConwayEra
+        -> Either Write.ErrInvalidTxOutInEra (Write.UTxO BabbageEra)
+    downgradeEra (Write.UTxO utxo) =
+        Write.utxoFromTxOutsInRecentEra
+        $ map (second Write.wrapTxOutInRecentEra)
+        $ Map.toList utxo
```

Comment:
Can you move these helper functions to `Internal.Cardano.Write.Tx`?

These function relate to the issue of converting `MaybeInRecentEra Write.UTxO` to a specific `Write.UTxO era`; their internal workings should not be a concern of `balanceTx`.

Notation: Instead of calling the function `assumeUTxOInEra`, I would call the function `forceUTxOInEra`, because the function can forcefully downgrade a Conway UTxO to a Babbage UTxO.

Reply:
> I would call the function `forceUTxOInEra`

In that case, would suggest changing "in" to "to", making it `forceUTxOToEra`.

Reply:
Moved

```haskell
forceUTxOToEra
    :: forall era1 era2. (IsRecentEra era1, IsRecentEra era2)
    => UTxO era1
    -> Either ErrInvalidTxOutInEra (UTxO era2)
forceUTxOToEra (UTxO utxo) =
    utxoFromTxOutsInRecentEra
    $ map (second wrapTxOutInRecentEra)
    $ Map.toList utxo
```

to `Write.Tx`, but left the dealing with the `MaybeInRecentEra` in a helper in `Cardano.Wallet.balanceTx`.